mlx-image (#1237)
huggingface/hub-docs: docs/hub/mlx-image.md

# Using mlxim at Hugging Face

[`mlxim`](https://github.com/riccardomusmeci/mlx-image) is an image models library built on Apple [MLX](https://github.com/ml-explore/mlx). It tries to replicate the great [timm](https://github.com/huggingface/pytorch-image-models) library from Ross Wightman, but for MLX models.

## Exploring MLX on the Hub

> You can find `mlxim` models by filtering at the left of the [models page](https://huggingface.co/models?library=mlxim&sort=trending).

Reviewer: ```suggestion
You can find `mlxim` models by filtering using the `mlxim` library name, like in [this query](https://huggingface.co/models?library=mlxim&sort=trending).
```

I believe the library name won't appear in the list of available libraries because it's being added with `filter=false`. We can revisit when the library gets more traction :)

Reviewer 2: Do you suggest keeping the sentence then?

Reviewer: Yes, for sure, we can keep the filter URL, just not mention the left column in the models page!

> There's also an open [mlx-vision](https://huggingface.co/mlx-vision) space for contributors converting and publishing weights for MLX format.

Reviewer: I don't see a way for people to join the community. Perhaps you want to consider these options? (taken from the settings of `mlx-community`)

<img width="703" alt="Screenshot 2024-03-11 at 10 47 29" src="https://github.com/huggingface/hub-docs/assets/1177582/64480970-7e3d-4f2b-a5c5-0903c350860f">

Reviewer 2: you were right.. now it's fixed.

Thanks to MLX Hugging Face Hub integration, you can load MLX models with a few lines of code. 

## Installation

```bash

> pip install mlx-image

Reviewer: This downgraded several of my packages, including PyTorch, which went from 2.1.2 to 2.0.0. Not related to this PR, but is that necessary?

Reviewer 2: No, it's not. I will fix it.

Reviewer: Thank you! 🙌 We can still progress with the docs and release them :)

```

> ## **Models**

Reviewer: ```suggestion

## Models

```

> Models weights are available on [`mlx-vision`](https://huggingface.co/mlx-vision) space on HuggingFace.

Reviewer: ```suggestion
Model weights are available on the [`mlx-vision`](https://huggingface.co/mlx-vision) community on HuggingFace.
```

"Space" could potentially be confused with [demo Spaces](https://hf.co/spaces)

> To create a model with weights:

Reviewer: ```suggestion
To load a model with pre-trained weights:

```

```python
from mlxim.model import create_model

# loading weights from HuggingFace

> model = create_model("resnet18")

Reviewer: Interesting! What repo are the weights coming from? Can we override?

Reviewer 2: Override HF weights is not possible currently, but it's a good point for the future. I currently assume to download weights from the defined mlx-vision hubs, but one can load the weights from his/her space. Working on it.

Reviewer: I didn't mean this was a requirement to release the docs! It's cool if you've already done it, otherwise we can just state where the weights are downloaded from.

# loading weights from local file

> model = create_model("resnet18", weights="path/to/weights.npz")

Reviewer: ```suggestion
model = create_model("resnet18", weights="path/to/resnet18/model.safetensors")
```

```

> To list all available models:

Reviewer: ```suggestion
To list all available models:

```

```python
from mlxim.model import list_models
list_models()
```

> > [!WARNING]
> > As of today (2024-03-08) mlx does not have `group` param for nn.Conv2d (e.g. `resnext`, `regnet`, `efficientnet`).

Reviewer: ```suggestion

> [!WARNING]
> As of today (2024-03-08) mlx does not support `group` param for nn.Conv2d. Therefore, architectures such as `resnext`, `regnet` or `efficientnet` are not yet supported in `mlxim`.
```

> ## **ImageNet-1K Results**

Reviewer: ```suggestion
## ImageNet-1K Results

```

I don't think bold is necessary in headings. Also, we usually use a blank line for separation to avoid potential rendering issues.

> Go to [results-imagenet-1k.csv](https://github.com/riccardomusmeci/mlx-image/blob/main/results/results-imagenet-1k.csv) to check every model converted and its performance on ImageNet-1K with different settings.

Reviewer: ```suggestion
Go to [results-imagenet-1k.csv](https://github.com/riccardomusmeci/mlx-image/blob/main/results/results-imagenet-1k.csv) to check every model converted to `mlxim` and its performance on ImageNet-1K with different settings.
```

Is there a tl;dr here? Like, performance is comparable with the original models?

> ## **Similarity to PyTorch and other familiar tools**
> `mlxim` tries to be as close as possible to PyTorch:

Reviewer: ```suggestion
## Similarity to PyTorch and other familiar tools

`mlxim` tries to be as close as possible to PyTorch:
```

> - `DataLoader` -> you can define your own `collate_fn` and also use `num_workers` to speed up the data loading

Reviewer: ```suggestion
- `DataLoader` -> you can define your own `collate_fn` and also use `num_workers` to speed up data loading
```

- `Dataset` -> `mlxim` already supports `LabelFolderDataset` (the good and old PyTorch `ImageFolder`) and `FolderDataset` (a generic folder with images in it)

> - `ModelCheckpoint` -> keeps track of the best model and saves it to disk (similar to PyTorchLightning) and it also suggests early stopping

Reviewer: ```suggestion
- `ModelCheckpoint` -> keeps track of the best model and saves it to disk (similar to PyTorchLightning). It also suggests early stopping.
```

> ## **Train**

Reviewer: ```suggestion
## Training

```

There's also an open [mlx-vision](https://huggingface.co/mlx-vision) space for contributors converting and publishing weights for MLX format.

Thanks to MLX Hugging Face Hub integration, you can load MLX models with a few lines of code. 

## Installation

```bash
pip install mlx-image
```

## Models

Model weights are available on the [`mlx-vision`](https://huggingface.co/mlx-vision) community on HuggingFace.

To load a model with pre-trained weights:
```python
from mlxim.model import create_model

# loading weights from HuggingFace (https://huggingface.co/mlx-vision/resnet18-mlxim)
model = create_model("resnet18") # pretrained weights loaded from HF

# loading weights from local file
model = create_model("resnet18", weights="path/to/resnet18/model.safetensors")
```

To list all available models:

```python
from mlxim.model import list_models
list_models()
```
> [!WARNING]
> As of today (2024-03-08) mlx does not support `group` param for nn.Conv2d. Therefore, architectures such as `resnext`, `regnet` or `efficientnet` are not yet supported in `mlxim`.

> ## ImageNet-1K Results

Reviewer: ```suggestion
## ImageNet-1K Results

```

Reviewer: final minor nit: it's best to use newlines before/after headings, just in case :)
